Add Image ID and Size Parameters to Azure node provider

psr-ai opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Currently there is no way to use private images within Azure's autoscaler. This is sometimes needed if we do not want to wait for the dependencies to be installed while spinning up new resources. Moreover, the disk size of the compute is also fixed. If the models are huge or multiple models are downloaded on the compute instances, it leads to no disk space. This PR adds both:

- `imageId` parameter to directly correspond to one of your Azure compute gallery images (use resource id)
- `osDiskSize` to control the size of the disk

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

psr-ai committed:
feat: configure image id parameter and add defaults to other image params

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
feat: add osDiskSize parameter for storageProfile

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

psr-ai committed:
chore: correction in the disk size description

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

psr-ai commented:
@edoakes looks like the ubuntu network is down while trying to fetch dependencies in the microcheck. Nothing related to my changes.

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

edoakes commented:
@psr-ai the Azure node provider is community maintained and the Ray maintainers (myself included) don't really have sufficient context to review it. @jjyao is finding the right person to review.

psr-ai commented:
@edoakes sounds good!

psr-ai commented:
Hi @edoakes @jjyao still waiting on this PR to be reviewed

jjyao commented:
Hi @psr-ai , thanks for the pinging. We are actively reaching out to Azure team to get an expert to review this PR.

jackfrancis commented on python/ray/autoscaler/_private/_azure/azure-vm-template.json:
```diff
@@ -227,16 +245,12 @@
                 "storageProfile": {
                     "osDisk": {
                         "createOption": "fromImage",
+			"diskSizeGB": "[parameters('osDiskSize')]",
```
If we introduce a static size of 128GB for existing users, won't that be a breaking change, depending on the VM SKU chosen? For example, if a VM SKU's built-in default is different than 128GB, now we are changing the VM configuration in a significant way. Especially, if a VM SKU's default size is larger than 128GB, we are now reducing the storage footprint, perhaps considerably.

We can obviously improve the docs and strongly recommend folks use this new configuration parameter, but we also need to accommodate for back-compat from the existing user community.

I'd recommend not declaring a default, and only including this new `osDiskSize` config if the user explicitly passes it in.

Also, should we update files like this with the new field?

- https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/azure/example-full.yaml

psr-ai committed:
feat: use diskSizeGB only if present else use defaults

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
feat: add examples of new parameters in example full

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai commented:
@jackfrancis I have addressed your comments:

- we had to specify default value as 0 for int (else we get an error to specify this value) and we use that to neglect the `diskSizeGB` by setting it `json(null)`
- have added commented out parameters on how to use them in the autoscaler file

Please let me know if anything else is needed!

psr-ai committed:
chore: format example full yaml file

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
feat: trim trailing whitespace

Signed-off-by: Prabhjot Singh Rai <hi@psr.ai>

psr-ai committed:
Merge branch 'master' into psr/azure-autoscaler/image-id-size

psr-ai commented:
Hi @jackfrancis, please rereview this PR, thank you!

jackfrancis reviewed (approved):
lgtm

tested this manually by deploying the ARM template: it's valid, also validated back-compat

jjyao reviewed (approved):
Unfortunately we currently don't have release test for cluster launcher on Azure. Relying on @jackfrancis 's manual test for now.

shubhbuz commented:
Can anyone confirm when these changes are going to be released ray python sdk
